sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -764,61 +768,202 @@ static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+ * one sdw link <---> one-cpu DAI <---> multi-codec DAIs
+ *
+ * multiple links with multiple slaves with aggregated = 1:
+ * link N <---> cpu DAIs N <----> codec DAIs I (I >= 1)
+ * link N <---> cpu DAIs M <----> codec DAIs J (J >= 1)
+ * ...
+ */
+static int get_slave_info(const struct snd_soc_acpi_link_adr *adr_link,
+			  int *cpu_dai_id, int *cpu_dai_num,
+			  int *codec_num, int *group_id,
+			  bool *group_generated)
+{
+	const struct snd_soc_acpi_adr_device *adr_d;
+	const struct snd_soc_acpi_link_adr *adr_next;
+	int index = 0;
+
+	*codec_num = adr_link->num_adr;
+	adr_d = adr_link->adr_d;
+
+	/* mask indicates only one sdw link */
+	if (!is_power_of_2(adr_link->mask))
+		return -EINVAL;
+
+	cpu_dai_id[index++] = ffs(adr_link->mask) - 1;
+	if (!adr_d->endpoints->aggregated) {
+		*cpu_dai_num = 1;
+		*group_id = 0;
+		return 0;
+	}
+
+	*group_id = adr_d->endpoints->group_id;
+
+	/* gather other link ID of slaves in the same group */
+	for (adr_next = adr_link + 1; adr_next && adr_next->num_adr;
+		adr_next++) {
+		const struct snd_soc_acpi_endpoint *endpoints;
+
+		endpoints = adr_next->adr_d->endpoints;
+		if (!endpoints->aggregated ||
+		    endpoints->group_id != *group_id)
+			continue;
+
+		if (!is_power_of_2(adr_next->mask))
+			return -EINVAL;
+
+		cpu_dai_id[index++] = ffs(adr_next->mask) - 1;
+		*codec_num += adr_next->num_adr;
+	}
+
+	/*
+	 * indicate CPU DAIs for this group have been generated
+	 * to avoid generating CPU DAIs for this group again.
+	 */
+	group_generated[*group_id] = true;
+	*cpu_dai_num = index;
+
+	return 0;
+}
+
+/* convert 1 -9 to '1' - '9' and A -F to 'A' - 'F' */
+static inline int int_to_ascii(int x)
+{
+	return x > 9 ? x + 87 : x + 48;
+}
+
+static int create_sdw_dailink(struct device *dev,
+			      struct snd_soc_dai_link *dai_links,
+			      int *be_id, int sdw_be_num, int sdw_cpu_dai_num,
+			      struct snd_soc_dai_link_component *cpus,
+			      const struct snd_soc_acpi_link_adr *link,
+			      int *cpus_id, bool *group_generated)
 {
-	struct snd_soc_dai_link_component *codec;
-	unsigned int part_id, link_id;
-	int i, j = 0, idx;
+	const struct snd_soc_acpi_link_adr *link_next;
+	struct snd_soc_dai_link_component *codecs;
+	int cpu_dai_id[SDW_MAX_DEVICES];
+	int cpu_dai_num, cpu_dai_index;
+	unsigned int part_id, group_id;
+	int codec_idx = 0;
+	int i, j = 0, k, idx;
+	int codec_num;
 	int ret;
 
-	codec = devm_kcalloc(dev, link->num_adr,
-			     sizeof(struct snd_soc_dai_link_component),
-			     GFP_KERNEL);
-	if (!codec)
+	ret = get_slave_info(link, cpu_dai_id, &cpu_dai_num, &codec_num,
+			     &group_id, group_generated);
+	if (ret)
+		return ret;
+
+	codecs = devm_kcalloc(dev, codec_num, sizeof(*codecs), GFP_KERNEL);
+	if (!codecs)
 		return -ENOMEM;
 
-	ret = create_codec_dai_name(dev, link, codec);
-	if (ret < 0)
-		return ret;
+	link_next = link;
+
+	/* generate codec name on different links in the same group */
+	for (i = 0; link_next && link_next->num_adr && i < cpu_dai_num;
```

Comment:
this is really begging for `for (link_next = link, i = 0;...)` Just as an idea, to make such loops with multiple instructions in start-stop-step statements I usually format them as
```
for (i = 0, a = a0, ...;
     i < N && a > amin &&...;
     i++, a--,...) {
```
but that's totally up to you, just an idea.

Reply:
yes, I like it. thanks!